File: demo/src/main/java/com/EMPhysics/Funcs.java
package com.EMPhysics;

import java.util.List;

public class Funcs {

    // Coulomb's law is a law describing the force between two charged particles
    public static double coulombLaw(ChargedParticle particle1, ChargedParticle particle2) {
        Vector2D direction = particle2.getPosition().subtract(particle1.getPosition());
        double distance = direction.magnitude();
        if (!checkDistanceValid(distance))
            return 0;
        double electricFieldMagnitude = (particle1.getCharge() * particle2.getCharge()
                * PhysicsConstants.COULOMB_CONSTANT) / (distance * distance);
        return electricFieldMagnitude;
    }

    // Calculating the force but with a vector
    public static Vector2D calculateNetElectricForce(ChargedParticle target, List<ChargedParticle> others) {
        Vector2D netForce = new Vector2D(0, 0);
        for (ChargedParticle other : others) {
            if (other == target)
                continue;

            Vector2D direction = other.getPosition().subtract(target.getPosition());
            double distance = direction.magnitude();
            if (!checkDistanceValid(distance))
                continue;

            double forceMagnitude = (target.getCharge() * other.getCharge() * PhysicsConstants.COULOMB_CONSTANT)
                    / (distance * distance);

            Vector2D forceVector = direction.normalize().scalar_multiply(forceMagnitude);
            netForce = netForce.add(forceVector);
        }
        return netForce;
    }

    public static double calculateElectricPotential(Vector2D point, ChargedParticle particle) {
        // particle's charge is the one creating the potential at point's position
        double distance = point.distanceTo(particle.getPosition());
        if (!checkDistanceValid(distance))
            return 0;
        return (particle.getCharge() * PhysicsConstants.COULOMB_CONSTANT) / distance;
    }

    public static Vector2D calculateNetElectricField(Vector2D point,
            List<ChargedParticle> particles) {
        Vector2D netElectricField = new Vector2D(0, 0);
        for (ChargedParticle p : particles) {
            Vector2D direction = point.subtract(p.getPosition());
            double distance = direction.magnitude();

            if (distance < PhysicsConstants.DISTANCE_DELTA)
                continue;

            double fieldMagnitude = (Math.abs(p.getCharge()) *
                    PhysicsConstants.COULOMB_CONSTANT)
                    / (distance * distance);

            Vector2D fieldVector = direction.normalize().scalar_multiply(fieldMagnitude);
            if (p.getCharge() < 0) {
                fieldVector = fieldVector.scalar_multiply(-1);
            }
            netElectricField = netElectricField.add(fieldVector);
        }
        return netElectricField;
    }

    public static boolean checkDistanceValid(double distance) {
        if (distance < PhysicsConstants.DISTANCE_DELTA) {
            return false;
        }
        return true;
    }
}

File: src/Physics/Funcs.ts
import { Vector2D } from './Vector2D';
import { ChargedParticle } from './ChargedParticle';
const constants = {
    k: 8.9875517873681764e9, // Coulomb's constant in N m²/C²
    epsilon_0: 8.854187817e-12, // Vacuum permittivity in C²/(N m²)
    e: 1.602176634e-19 // Elementary charge in C
}
export function coulombLaw(p1: ChargedParticle, p2: ChargedParticle): number {
    const distance: number = p2.getPosition().subtract(p1.getPosition()).magnitude();
    const forceMagnitude = (constants.k * p1.getCharge() * p2.getCharge()) / (distance * distance);
    return forceMagnitude;
}


export function calculateNetElectricForce(targetParticle: ChargedParticle, otherParticles: ChargedParticle[]): Vector2D {
    let netField = new Vector2D(0, 0);
    for (const particle of otherParticles) {
        if (particle === targetParticle) continue;

        const direction: Vector2D = targetParticle.getPosition().subtract(particle.getPosition());
        const distance: number = direction.magnitude();
        if (distance < 1e-6) continue;
        const unit = direction.normalize();
        const magnitude = (constants.k * targetParticle.getCharge() * particle.getCharge()) / (distance * distance);

        netField = netField.add(unit.scalar_multiply(magnitude));
    }
    return netField;
}

export function calculateElectricPotential(targetParticle: ChargedParticle, otherParticles: ChargedParticle[]): number {
    let potential = 0;
    for (const particle of otherParticles) {
        if (particle === targetParticle) continue;

        const distance: number = particle.getPosition().subtract(targetParticle.getPosition()).magnitude();
        potential += (constants.k * particle.getCharge()) / distance;
    }
    return potential;
}

export function calculateNetElectricField(point: Vector2D, otherParticles: ChargedParticle[]): Vector2D {
    let netField = new Vector2D(0, 0);
    for (const particle of otherParticles) {
        const direction = point.subtract(particle.getPosition());
        const distance = direction.magnitude();
        if (distance < 1e-6) continue;

        const unit = direction.normalize();
        const magnitude = (constants.k * particle.getCharge()) / (distance * distance);

        netField = netField.add(unit.scalar_multiply(magnitude));
    }
    return netField;
}